channeldb/invoices.go

```diff
@@ -1876,60 +1881,48 @@ func settleHtlcsAmp(invoice *invpkg.Invoice,
 
 // updateInvoice fetches the invoice, obtains the update descriptor from the
 // callback and applies the updates in a single db transaction.
-func (d *DB) updateInvoice(hash *lntypes.Hash, refSetID *invpkg.SetID, invoices,
-	settleIndex, setIDIndex kvdb.RwBucket, invoiceNum []byte,
-	callback invpkg.InvoiceUpdateCallback) (*invpkg.Invoice, error) {
-
-	// If the set ID is non-nil, then we'll use that to filter out the
-	// HTLCs for AMP invoice so we don't need to read them all out to
-	// satisfy the invoice callback below. If it's nil, then we pass in the
-	// zero set ID which means no HTLCs will be read out.
-	var invSetID invpkg.SetID
-	if refSetID != nil {
-		invSetID = *refSetID
-	}
-	invoice, err := fetchInvoice(invoiceNum, invoices, &invSetID)
-	if err != nil {
-		return nil, err
-	}
+func (d *DB) updateInvoice(hash *lntypes.Hash, invoices,
+	settleIndex, setIDIndex kvdb.RwBucket, invoice *invpkg.Invoice,
+	invoiceNum []byte, callback invpkg.InvoiceUpdateCallback) (
+	*invpkg.Invoice, error) {
 
 	// Create deep copy to prevent any accidental modification in the
 	// callback.
-	invoiceCopy, err := invpkg.CopyInvoice(&invoice)
+	invoiceCopy, err := invpkg.CopyInvoice(invoice)
 	if err != nil {
 		return nil, err
 	}
 
 	// Call the callback and obtain the update descriptor.
 	update, err := callback(invoiceCopy)
 	if err != nil {
-		return &invoice, err
+		return invoice, err
 	}
 
 	// If there is nothing to update, return early.
 	if update == nil {
-		return &invoice, nil
+		return invoice, nil
 	}
 
 	switch update.UpdateType {
 	case invpkg.CancelHTLCsUpdate:
-		return d.cancelHTLCs(invoices, invoiceNum, &invoice, update)
+		return d.cancelHTLCs(invoices, invoiceNum, invoice, update)
```

Comment:
I think the enum above is part of the way there, but rather then switch on the enum, we make a new `InvoiceUpdater` from the invoice itself (the KV impl maybe still copies like above), then we do `ApplyUpdate` on each of the abstract update descs, which are the items in the callback today), then finally commit. The internal impl then does a type switch or w/e on each of the updates to figure out how to apply it.

Reply:
I choose to go with a similar although quite different approach. The update desc and in-memory update stays the same to avoid very intrusive code changes and instead the in-memory stuff is restructured a bit so that the DB stuff can be more fine-grained (which is ideal for the SQL stuff). PTAL all the commits, especially the one adding the `InvoiceUpdater` which will clear up why these changes are made.